src/Compilers/Core/Portable/Syntax/GreenNode.NodeEnumerable.cs

```diff
@@ -0,0 +1,72 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Microsoft.CodeAnalysis.PooledObjects;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis;
+
+internal abstract partial class GreenNode
+{
+    [NonCopyable]
+    public ref struct NodeEnumerable(GreenNode node)
+    {
+        public readonly Enumerator GetEnumerator()
+            => new Enumerator(node);
+
+        [NonCopyable]
+        public ref struct Enumerator
+        {
+            private readonly ArrayBuilder<Syntax.InternalSyntax.ChildSyntaxList.Enumerator> _stack;
+
+            private bool _started;
+            private GreenNode _current;
+
+            public Enumerator(GreenNode node)
+            {
+                _current = node;
+                _stack = ArrayBuilder<Syntax.InternalSyntax.ChildSyntaxList.Enumerator>.GetInstance();
+                _stack.Push(node.ChildNodesAndTokens().GetEnumerator());
+            }
+
+            public readonly void Dispose()
+                => _stack.Free();
+
+            public readonly GreenNode Current
+                => _current;
```

Comment:
Can we assert that you don't access `_current` until `_started` is set?

Reply:
sure!  though, generally, `yield` doesn't do that, and our own docs say such things are undefined.  But i don't mind adding guardrails :)